Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-    ///
-    /// This specialisation (for UInt) is faster than the generic BinaryIntegers implementation (earlier in this file).  It is used as an opportunistic fast-path in the generic implementation, for any values that happen to fit into a UInt.
-    internal var numericStringRepresentation: ArraySlice<UInt8> {
-        // This property's type is an ArraySlice rather than a straight ContiguousArray because it's computationally more expensive to size the array exactly right (whether by pre-calculating the required size more precisely, or by dynamically resizing it during execution of the iterative algorithm).  Since this is intended to only be a transient value, for the purposes of translating from BinaryInteger to the ICU libraries, this is the right compromise of runtime speed (CPU time) and memory efficiency (usually just one excess byte, if any).
-
-        // It's easier to just special-case zero than handle it in the main algorithm.
-        guard .zero != self else {
-            return [UInt8(ascii: "0")]
-        }
-
-        let maximumDigits = Self.maximumDecimalDigitsForUnsigned(bitWidth: self.magnitudeBitWidth)
-        var actualDigits: Int = Int.min // Actually initialised inside the closure below, but the compiler mistakenly demands a default value anyway.
-
-        return ContiguousArray(unsafeUninitializedCapacity: maximumDigits) { buffer, initialisedCount in
-            actualDigits = numericStringRepresentation(intoEndOfBuffer: &buffer[...])
-
-            let unusedDigits = maximumDigits - actualDigits
-            assert(0 <= unusedDigits, "Negative unused digits \(unusedDigits)!  Expected at most \(maximumDigits) digit(s) but emitted \(actualDigits).")
-
-            if 0 < unusedDigits {
-                buffer[0..<unusedDigits].initialize(repeating: 0) // The buffer is permitted to be partially uninitialised, but only at the end.  So we have to initialise the unused portion at the start, in principle.  Technically this probably doesn't matter given we never subsequently read this part of the buffer, but there's no way to express that such that the compiler can ensure it for us.
-            }
-
-            initialisedCount = maximumDigits
-        }[(maximumDigits - actualDigits)...]
+/// Returns the largest `exponent` and `power` in `pow(10, exponent) <= UInt.max + 1`.
+///
+/// The `exponent` is also the maximum number of decimal digits needed to represent a binary integer
+/// in the range of `0 ..< power`. Another method is used to estimate the total number of digits, however.
+/// This is so that binary integers can be rabased and encoded in the same loop.
+///
+/// ```
+/// 32-bit: (exponent:  9, power:           1000000000)
+/// 64-bit: (exponent: 19, power: 10000000000000000000)
+/// ```
+///
+private func maxDecimalExponentAndPowerForUnsignedIntegerWord() -> (exponent: Int, power: UInt) {
```

Comment:
As a sidenote, this gets evaluated at compile time and reduced down to the actual constants, which is pretty awesome.  Its use later becomes two inline instructions, to load the immediates 0x8ac7230489e80000 and 0x13 (on x86-64).

Reply:
Yeah, I assume it's fine to rely on this compiler optimisation.  Plus I spent quite a while trying to come up with _any_ alternative that's remotely scalable and elegant, but found nothing.

That's also why in my earlier implementation I specifically commented about that, in the code.  I just hadn't discovered, at that point, that the compiler _does_ in fact optimise it, so I believe my comment was along the lines of "I _hope_ the compiler can optimise this…". 😁